components/rich_text_editor/rich_text_editor.tech_spec.md

```diff
@@ -292,27 +332,198 @@ feed items also. The mentions/channels list dropdowns can probably be purpose
 The extension should handle inputting or "typing" an emoji and pasting one from
 clipboard. To add functionality for searching emojis by typing `:` and some text
 can be a stretch goal, but it depends on getting the emoji picker into Dialtone
 first.
 
-### Codeblocks
+## Codeblocks
 
 We can customize the
 [extension-code-block](https://tiptap.dev/api/nodes/code-block) package from
 TipTap. Similar to above, the actual node component is a separate Dialtone
 component that we'll have to pass to the extension.
 
-### Links
+## Links
 
 We'll build a custom extension that recognizes URLs typed/pasted into the editor
 and makes them into actual HTML links. Does not allow linkifying non-URL text or
 adjusting any attributes on the `<a>` tag itself, but the purpose is more to be
-a visual cue to show that an URL was typed.
+a visual cue to show that an URL was typed. Therefore the links are not
+clickable either, but they should behave as normal text – just with a styling to
+look like links.
+
+### Overview
+
+After a brief [investigation](https://dialpad.atlassian.net/browse/DT-1153) and
+some POC work it became apparent that we can't unfortunately rely on the
+`@tiptap/extension-link` package that provides similar functionality. The
+package wouldn't allow us to customize the behavior of the extension to meet our
+needs, so we'll have to build a custom extension instead. It would for example
+only validate the link after a following whitespace character, which would leave
+all the links at the end of a message not linkified. With a custom extension we
+can create rules to linkify text as it is being typed.
+
+Also worth noting that with my current understanding we should construct this as
+a Mark instead of a Node. For example backspacing at the beginning of a Mark and
+merging its contents with the preceeding content seems easier than with a Node,
+where the default behavior is to exit and remove the Node and convert its
+contents back to plain text. The main consequence of this decision is that Marks
+can render only native HTML tags, so in this case an `<a>` tag with `d-link`
+class instead of `DtLink`. To the user they would look indistinguishable though
+and since the links are not operational this should not be a big deal.
+
+### Scope
+
+We should linkify the same type of content that Dialpad backend does, so the
+following:
+
+- URLs and IPs
+  - Without a protocol, e.g. `dialpad.com` or `127.0.0.1`
+  - With path(s), e.g. `dialpad.com/news` or `127.0.0.1/news`
+  - With param(s), e.g. `dialpad.com?cache=1` or `127.0.0.1?cache=1`
+  - With a protocol, e.g. `https://dialpad.com`
+- Email Addresses
+  - Without `mailto:` prefix, e.g. `mika@dialpad.com`
+  - With `mailto:` prefix, e.g. `mailto:mika@dialpad.com`
+  - With param(s), e.g. `mika@dialpad.com?subject=hey`
+- Phone Numbers
+  - 7-15 characters long
+  - Without separators, e.g. `7144107035`
+  - With dashes `-`, e.g. `714-410-7035`
+  - With whitespace `\s`, e.g. `714 410 7035`
```

Comment:
I feel like "with whitespace" might end up causing us grief...just a note that if it ends up being a hassle or performance issue, I would be ok NOT highlighting this format in the editor